### DISCUSSION TITLE
Force CamelCase for class names and snake_case for function names

### DISCUSSION THREAD

**Discussion by sylee957** [sympy/sympy#24569] [General]
I think that most of the bad examples that doesn't follow PEP-8 are from by `sympy.functions` and `sympy.stats`, 
because people have `CamelCase` and `snake_case` oppositely before, and it is more complicated now because there have been exceptions (things like `Abs, Min, Max`) that starts to bring more inconsistency, and more people find the old bad practices as 'standards' and keep adding stuff in that way.

https://github.com/sympy/sympy/blob/c25ecf0ef7c63ec8b9cb416544a682c758fcc5f2/sympy/functions/__init__.py#L8-L49
https://github.com/sympy/sympy/blob/c25ecf0ef7c63ec8b9cb416544a682c758fcc5f2/sympy/stats/__init__.py#L160-L194

I think that it is possible to obey this if we can simply make everything as:

```python3
class Sin(Function):
    ...

def sin(x):
    return Sin(x)
```

Probably we can make something like a huge naming alias stub module, like `__future__` where users are encouraged to pick the right function/class definitions they are suitable, but we could have consistency in the library.

Besides, I think that all the constructor and AST should be split (such as `__new__` or `Function.eval`) could move to functions that are trivially defined like above, and make `__new__` more lightweight. 
I think that it is impossible for sympy to construct very huge terms because of performance issues, about it being busy about computations in the constructor. If it is not needed, it could be users to choose not to eagerly evaluate terms at all, without a need to propagate `evaluate=False`

**Comment by asmeurer** [comment]
This sort of thing was discussed before, mostly with the idea of splitting the core classes as add()/Add() where the latter would do no evaluation. I generally agree that the SymPy classes should do minimal or even no evaluation. Whether this also means there should be a corresponding lowercase function that evaluates is something I'm not sure about. Most evaluation can be achieved with a method, like doit(), which has the benefit of applying recursively to an entire expression. Maybe for add() and mul() it would make sense because there's definitely some minimal evaluation (i.e., cancellation) you almost always want there, which could be distinguished from theoretical Add() and Mul() classes that do no evaluation at all. For other functions, like sin, typically the only automatic evaluation that you'd want is evaluation on explicit numbers (like `sin(pi)`). Right now the performance problems come from trying to do way more than that (especially things that make use of assumptions in constructors). 

Generally I do think that end user code is going to want to write mathematical functions using common mathematical conventions rather than PEP 8 conventions. It's also much easier to type code when it's all lowercase. The only real benefit that PEP 8 class vs. function naming would bring to the existing SymPy functions is it would make it clearer which functions can be unevaluated. But some functions which currently can't be unevaluated should really be converted to classes so they can be, and we wouldn't want to have to change their name just to do that.

**Reply by sylee957** [reply]
> Generally I do think that end user code is going to want to write mathematical functions using common mathematical conventions rather than PEP 8 conventions

Well, it is not obvious beyond a couple of simple examples like `sin, cos`. Nobody does math with `sqrt, factorial, bessel_j, Integral` explicitly written down as plaintexts. What we actually call 'mathematical' with pragmatic notation is already full of naming workaround.
